src/accelerate/utils/numactl.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import subprocess
+import sys
+
+from accelerate.utils.environment import _get_nvidia_smi
+
+
+def get_query_command(device_id: int = None, backend: str = None):
+    """
+    Returns the right param args for `numa` support based on a `backend`.
+
+    Currently only supports `nvidia`.
+    """
+    if backend == "nvidia":
+        return [_get_nvidia_smi(), "--query-gpu=pci.bus_id", "-i", str(device_id), "--format=csv,noheader"]
+    else:
+        # TODO: Add support for AMD and Gaudi2
+        raise ValueError(f"Backend {backend} is currently not supported through `accelerate launch`.")
+
+
+def get_bus_id(device_id: int = None, backend: str = None):
+    """
+    Gets the bus ID for `device_id`. If not passed, will default to the local rank.
+    """
+    if device_id is None:
+        device_id = os.environ.get("LOCAL_RANK", 0)
+    command = get_query_command(device_id, backend)
+    output = subprocess.check_output(command, universal_newlines=True)
+    return output.strip()[4:]
+
+
+def get_numa_node_for_device(device_id: int = None):
+    """
+    Gets the numa node for `device_id`. If not passed, will default to the local rank.
+    """
+    bus_id = get_bus_id(device_id)
+    if bus_id is None:
+        bus_id = 4
+    return subprocess.check_output(["cat", f"/sys/bus/pci/devices/{bus_id}/numa_node"]).strip().decode()
+
+
+def launch():
+    args = sys.argv
+    script, *script_args = args[1:]
+    node = get_numa_node_for_device()
+    cmd = ["numactl", f"--cpunodebind={node}", f"--membind={node}", "python3", script] + script_args
+    process = subprocess.Popen(cmd, env=os.environ.copy())
+    process.wait()
+    if process.returncode != 0:
+        raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
```

Comment:
why do you need to do that when you can do it from python?

https://github.com/NVIDIA/DeepLearningExamples/blob/9dd9fcb98f56187e49c5ee280cf8dbd530dde57b/TensorFlow2/LanguageModeling/BERT/gpu_affinity.py#L55-L63

This is the pynvml way I shared in the OP of the issue, but we are just removing `pynvml`

Reply:
I also didn't see where you check that `numactl` exists - it is not installed on linux by default. So doing it via python is probably much simpler.

Reply:
@stas00 do you know if we would be okay doing this in the `AcceleratorState`?

Or better phrased as:

Is it paramount that this gets set before things like `torch.distributed` is initialized

Reply:
yes, of course, you can change the affinity dynamically after the process started:

Quote from https://linux.die.net/man/2/sched_setaffinity
> If the process specified by pid is not currently running on one of the CPUs specified in mask, then that process is migrated to one of the CPUs specified in mask. 

i.e. it doesn't matter how the process is started.